opsdroid/events.py

```diff
@@ -0,0 +1,166 @@
+        connector: Connector object used to interact with given chat service
+        text: Text of message as string
+        raw_message: Raw message provided by chat service
+        regex: A re match object for the regular expression message was matched
+            against
+        responded_to: Boolean initialized as False. True if message has been
+            responded to
+
+    """
+
+    def __init__(self, user, room, connector,
+                 text, raw_message=None):  # noqa: D107
+        """Create object with minimum properties."""
+        super().__init__(user, room, connector)
+        self.text = text
+        self.raw_message = raw_message
+        self.regex = None
+
+    async def _thinking_delay(self):
+        """Make opsdroid wait x-seconds before responding.
+
+        Number of seconds defined in YAML config. file, accessed via connector.
+        """
+        seconds = self.connector.configuration.get('thinking-delay', 0)
+
+        if isinstance(seconds, list):
+            seconds = randrange(seconds[0], seconds[1])
+
+        await asyncio.sleep(seconds)
+
+    async def _typing_delay(self, text):
+        """Delays reply to simulate typing.
+
+        Seconds to delay equals number of characters in response multiplied by
+        number of seconds defined in YAML config. file, accessed via connector.
+        """
+        seconds = self.connector.configuration.get('typing-delay', 0)
+        char_count = len(text)
+
+        if isinstance(seconds, list):
+            seconds = randrange(seconds[0], seconds[1])
+
+        await asyncio.sleep(char_count*seconds)
+
+    async def respond(self, response_event, room=None):
+        """Respond to this message using the connector it was created by.
+
+        Creates copy of this message with updated text as response.
+        Delays message if thinking or typing delay present in config. file.
+        Updates responded_to attribute to True if False.
+        Logs response and response time in OpsDroid object stats.
+        """
+        opsdroid = get_opsdroid()
+        if isinstance(response_event, str):
+            response = copy(self)
+            response.text = response_event
+        else:
+            response = response_event
+
+        if 'thinking-delay' in self.connector.configuration or \
+           'typing-delay' in self.connector.configuration:
+            await self._thinking_delay()
+            await self._typing_delay(response.text)
+
+        await self.connector.respond(response, room)
+        if not self.responded_to:
+            now = datetime.now()
+            opsdroid.stats["total_responses"] = \
+                opsdroid.stats["total_responses"] + 1
+            opsdroid.stats["total_response_time"] = \
+                opsdroid.stats["total_response_time"] + \
+                (now - self.created).total_seconds()
+            self.responded_to = True
+
+    async def react(self, emoji):
```

Comment:
This should emit a `Reaction` event!!

Reply:
Arguably, this whole method should be deprecated in favour of some type of call to `Message.send` i.e.:

```python
Message.send(Reaction("😼"))
```

Reply:
I agree on both points, but I think it's worth leaving for now and putting in the next PR.

Reply:
Agree and agree that it can wait until the next PR.